### DISCUSSION TITLE
Concurrency support

### DISCUSSION THREAD

**Discussion by seanmrich** [stephencelis/swift-gen#29] [Q&A]
Building `swift-gen` with swift 6 raises diagnostics around the static declarations. For example:
```swift
extension Gen where Value == Bool {
  /// A generator of random boolean values.
  public static let bool = Gen { rng in Bool.random(using: &rng) }
  //               👆
  // Static property 'bool' is not concurrency-safe because non-'Sendable' type 'Gen<Bool>' may have shared mutable state
}
```

The problem of course is that `Gen` isn't `Sendable`. I was going to put together a pull request to add support, but it isn't clear what the right direction would be here. `Gen._run` would need to be a `Sendable` closure, but it seems too high a requirement to require this for all users of the type. I'm not aware of a way to define `Gen` to be `Sendable` only when its closure is `Sendable`. Without that, perhaps these static declarations need to be computed properties (at least when in swift 6 language mode)?

**Comment by stephencelis** [comment]
@seanmrich I think a pull request requiring `@Sendable` is totally reasonable for the library as currently designed.

In the future we would probably rethink `Gen` as a protocol hierarchy, which would allow for sendability (and even hashability) at the conformance level, but we're not sure if/when we'll revisit that.

**Reply by seanmrich** [reply]
I can work on a PR for that, but it would require a tweak. We'd run into the same issue as in my [earlier PR](https://github.com/pointfreeco/swift-gen/pull/28) since the CI is using swift 5.3. The package's tools version would have to require at least swift 5.5 (maybe even more recent -- can't remember when the logic around static properties was introduced). Not sure if you have time to update to a more modern version.

**Reply by mbrandonw** [reply]
@seanmrich You can bump the package Swift version all the way up to 5.9 if it makes things easier. Most of our packages are 5.9+ already. Or if you want to keep things a little more backwards compatible 5.8 is also possible, but I don't think we need to go any earlier than that.

**Reply by stephencelis** [reply]
@seanmrich And sorry if I wasn't clearer in my comment, but I was suggesting bumping the CI to a newer version. Here's an example:

https://github.com/pointfreeco/swift-case-paths/blob/main/.github/workflows/ci.yml#L51-L64

**Reply by seanmrich** [reply]
That's helpful thanks. Never really used those workflows so the example is very helpful. Let me see what I can do

**Reply by seanmrich** [reply]
[Pull request](https://github.com/pointfreeco/swift-gen/pull/30) submitted. I did try simplifying the many `zip` functions using parameter packs like this:
```swift
func zip<each A: Sendable>(_ element: repeat Gen<each A>) -> Gen<(repeat each A)> {
  return Gen { rng in
    (repeat (each element)._run(&rng))
  }
}
```
Unfortunately the compiler thinks `repeat Gen<each A>` isn't `Sendable`. There's already an [issue](https://github.com/swiftlang/swift/issues/74392) submitted.
